Add support of sitemap generation

raspopov opened the pull request:
Fixes [#35523](https://mantisbt.org/bugs/view.php?id=35523).

Discussion on [forum](https://mantisbt.org/forums/viewtopic.php?p=81300).

raspopov committed:
Add support of sitemap generation.

Fixes #35523.

vboctor commented on sitemap.php:
```diff
@@ -0,0 +1,102 @@
+$t_threshold = config_get( 'view_bug_threshold' );
+
+# Maximum number of URLs
+$t_max_urls = 50000;
```
Use `$g_`

raspopov replied:
Thanks for your interest. But this variable is not a global setting, just a constant for the code below, put in the file header to make it easier to find. I don't think the standard will change, and 50000 will never change.

vboctor commented on sitemap.php:
```diff
@@ -0,0 +1,102 @@
+# Maximum number of URLs
+$t_max_urls = 50000;
+
+$t_path = config_get_global( 'path' );
```
Use `$g_`

vboctor commented on core/layout_api.php:
```diff
@@ -91,18 +93,26 @@ function layout_page_header_begin( $p_page_title = '' ) {
-	$t_title = htmlspecialchars( config_get( 'search_title' ) );
-	$t_searches = array( 'text', 'id' );
-	foreach( $t_searches as $t_type ) {
+	if( auth_anonymous_enabled() ) {
```
Should you use `current_user_is_anonymous()` instead? Is the sitemap useful for signed in users?

raspopov replied:
>  Is the sitemap useful for signed in users?

Yes, sitemap is for search engines only. They'll never log in anywhere.

vboctor commented on core/layout_api.php:
```diff
@@ -91,18 +93,26 @@ function layout_page_header_begin( $p_page_title = '' ) {
 			"\n";
 	}
 
+	# Advertise the availability of the browser search plug-ins.
```
Is this a different feature?

raspopov replied:
Good point, I'm trying to stick small changes together again... 😔 Moved to #2126.

vboctor commented on sitemap.php:
```diff
@@ -0,0 +1,102 @@
+echo '<?xml version="1.0" encoding="UTF-8"?>', "\n",
+	'<urlset xmlns="http://www.sitemaps.org/schemas/sitemap/0.9">', "\n";
+
+if( auth_anonymous_enabled() ) {
```
Should you use `current_user_is_anonymous()` instead?

raspopov replied:
The current_user_is_anonymous() function not only uses the auth_anonymous_enabled() function, but also checks cookies in the database, which seems unnecessary in this situation.

vboctor reviewed (changes requested):
I think we should try to reduce the amount of work that is done from anonymous end points when we can. How about we have this script generate a sitemap cache file that is re-used and only refreshed if older than an hour.

vboctor commented on sitemap.php:
```diff
@@ -0,0 +1,102 @@
+$t_threshold = config_get( 'view_bug_threshold' );
+
+# Maximum number of URLs
+$t_max_urls = 50000;
```
If the generation of the sitemap is limited to once an hour, then we can remove the max limit which is arbitrary and not configurable.

raspopov replied:
Search engines load sitemaps, if they load them at all, at best once before a site's re-indexing, once a day or less often.

raspopov commented:
> I think we should try to reduce the amount of work that is done from anonymous end points when we can.

Search engines load sitemaps, if they load them at all, at best once before a site's re-indexing, once a day or less often.

raspopov commented:
If #2127 is accepted, I will be able to add more links to the sitemap.

raspopov committed:
Add Sitemap test.

Strict schema validation.

raspopov committed:
Fix current project id.

raspopov committed:
Add news links and more tests.

The sitemap.php file has been modified to make it test-friendly.
Unnecessary sorting of issues has been removed.

The tests are now run using GuzzleHttp\Client.
More tests have been added.

raspopov committed:
Add error diagnostic message.

raspopov committed:
Compact the code

An unnamed class was used and the test namespace was restored.
